### PR TITLE
(CDAP-1402) Refactor stream admin and stream coordinator

### PR DESCRIPTION
**Pull request by chtyim**
- Use lock to simplify coordination logic
- Remove circular dependencies between FileStreamAdmin and AbstractStreamCoordinatorClient

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinatorClient.java [hunk 1]
```diff
@@ -53,30 +39,23 @@
 public abstract class AbstractStreamCoordinatorClient extends AbstractIdleService implements StreamCoordinatorClient {
 
   private static final Logger LOG = LoggerFactory.getLogger(AbstractStreamCoordinatorClient.class);
-  private static final Gson GSON = new GsonBuilder()
-    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
-    .create();
+  private static final Gson GSON = new Gson();
```
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinatorClient.java]: If we're using this GSON instance to serialize/deserialize ``CoordinatorStreamProperties``, which contain a ``Schema`` - via the ``Formatspecification`` class, shouldn't we leave the ``SchemaTypeAdapter`` here?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 2]
```diff
@@ -347,25 +337,21 @@ private StreamConfig getAndValidateConfig(StreamConfig currConfig, HttpRequest r
...
+    // Validate ttl
+    Long ttl = properties.getTTL();
+    if (ttl != null) {
+      if (ttl < 0) {
         responder.sendString(HttpResponseStatus.BAD_REQUEST, "TTL value should be positive.");
         return null;
       }
       // TTL in the REST API is in seconds. Convert it to ms for the config.
-      newTTL = TimeUnit.SECONDS.toMillis(newTTL);
+      ttl = TimeUnit.SECONDS.toMillis(ttl);
```
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: Why not do this operation in the ``StreamPropertiesAdapter.deserialize`` instead?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinatorClient.java (hunk 1 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinatorClient.java]: That's right. Need that adapter here.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java (hunk 2 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: This is from the old code. Probably that's the place of inconsistency about converting the unit of ttl. Will take a look into that.

**Comment by jguery** [comment]
LGTM :+1:

**Comment by anwar6953** [comment]
You changed the PUT /streams/{streamName}/config method to expect a JSON-encoded StreamProperties object now, instead of a StreamConfig.
(So now, its consistent with the /info endpoint). Should it be /streams/{streamName}/properties instead?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java [hunk 3]
```diff
@@ -125,4 +115,43 @@ public String toString() {
...
+      this.formatSpec = formatSpec;
+    }
+
+    public void setNotificationThreshold(int notificationThreshold) {
+      this.notificationThreshold = notificationThreshold;
+    }
+
+    public StreamConfig build() {
+      return new StreamConfig(config.getName(), config.getPartitionDuration(), config.getIndexInterval(),
+                              Objects.firstNonNull(ttl, config.getTTL()),
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java]: Why is this needed? It looks like ttl will never be null.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java [hunk 4]
```diff
@@ -125,4 +115,43 @@ public String toString() {
...
+                              config.getLocation(),
+                              Objects.firstNonNull(formatSpec, config.getFormat()),
+                              Objects.firstNonNull(notificationThreshold, config.getNotificationThresholdMB()));
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java]: also here. It looks like notificationThreshold can never be null.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java (hunk 3 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java]: nevermind.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java (hunk 4 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/StreamConfig.java]: nevermind.

**Comment by anwar6953** [comment]
Build: https://builds.cask.co/browse/CDAP-DUT700-14

**Comment by chtyim** [comment]
The put config method always expect a json object that can have any of the TTL, threshold and format set. How the server side parse it is independent of the rest endpoint. From that point of view, the streamproperties is more appropriate because it use intended for server client communication, while the streamconfig is purely internal to the stream system

**Comment by chtyim** [comment]
Build passed. Merging changes

**Comment by anwar6953** [comment]
Right, my point was just to bring up whether or not the endpoint should be renamed
